README.md
# 22 July 2014 modified by Trey Wenger
# 19 Dec 2015 modified by GeeksBsmrT
# Introduction

This is for Ingress. If you don't know what that is, you're lost.

This code is designed to make a plan for linking a given set of portals in the
way (and the order) that creates the most fields. This is harder than it sounds.
If you're working on more than a dozen portals, learning to use this code may
be faster than planning by hand.

This code follows the procedure in Jonathan Baker's [YouTube video][0].


# Prerequisites

You'll need [Python3][1] as well as networkx, numpy, and matplotlib.

# Example

I'll be distributing this code with a file Test.csv. Try running

    python3 makePlan.py -n 4 Test.csv

This will put a bunch of files into the "%USRDIR%/Ingress/Fielding/%inputFileName%" directory (see OUTPUT FILE LIST)

Now try running

    python3 makePlan.py -n 3 "%USRDIR%/Ingress/Fielding/%inputFileName%/%inputFileName%.plk"

This uses the plan stored in output.pkl instead of calculating a new one. It will create files for 3 agents instead of 4.

### OUTPUT FILE LIST

	keyPrep.txt
		List of portals, their numbers on the map, and how many keys are needed

	keys_for_agent_M_of_N.txt
		List of keys agent number M will need (if N agents are participating)

	links_for_agent_M_of_N.txt
		List of ALL the links
		Total distance traveled and AP earned by agent number M
			* Except for the links marked with a star (*), the links should be made IN THE ORDER LISTED
			* Links with a star can be made out of order, but only EARLY i.e. BEFORE their position in the list (this can save you time)
			* The links that agent number M makes are marked with underscores__
			* The first portal listed is the origin portal (where the agent must be)
			* The second portal listed is the destination portal (for which the agent must have a key)

	portalMap.png
		A map showing the locations of the portals
	linkMap.png
		A map showing the locations of portals and links
			* Up is north
			* Portal numbers increase from north to south
			* Portal numbers match "keyPrep.txt" and "linkes_for_agent_M_of_N.txt"
			* Link numbers match those in the link schedules "links_for_agent_M_of_N.txt"

	ownershipPrep.txt
		List of portals whose first link is incoming
			* These portals need to be captured and fully powered before the linking operation
		List of portals whose first link is outgoing
			* You may be able to save time by capturing and fully powering these portals DURING the linking operation

	lastPlan.pkl
		A Python pickle file containing all portal and plan information
			* The default name is "lastPlan.pkl"
			* In the examples above, this is called "output.pkl"

# Warranty

No promises

# Usage

    python3 makePlan.py -n agent_count input_file

    agent_count: Number of agents for which to make a plan

    input_file:  One of two types of files:
        .csv   format:
PORTAL NAME, INTEL MAP LINK, (OPTIONAL:) NUMBER OF KEYS AVAILABLE

Example:
 Catholic Church of the Holy Comforter, https://www.ingress.com/intel?ll=38.031745,-78.478592&z=18&pll=38.031796,-78.479439, 3

            portal name should not contain commas
            keys (optional parameter) is the number of keys you have for the portal
            If you leave this blank, the program assumes you have no keys

        .pkl   an output from a previous run of this program
            this can be used to make the same plan with a different number of agents

# Notes

The space of possible max-field plans is large. Rather than trying every
possibility, this program randomly tries some plans and presents you with one
that doesn't require you to obtain too many more keys.

If you don't like the plan you got, run it again. You'll probably get a
different plan.


[0]: https://www.youtube.com/watch?v=priezq6Dm4Y
[1]: https://www.python.org/


changelog
Changes: 19 Dec 2015 - GeeksBsmrt V3.0
1. Upgraded for Python 3
2. Changed colored output to Resistance Blue
3. Removed the requirement for output folder/file names
	3a. Output Folder/File will be created based on the Input filename, and placed in subfolder of %USRDIR%/Ingress/Fielding
4. Automatically uses Google Maps API for output files
5. Increased max portals from 50 to 1000

==========================================================================
Changes: 15 August - 
1. Fixed bug in Optimization criteria
2. Increased number of optimization attempts
3. Forced limitation on number of portals
4. Implemented google maps API on portal map and link map
5. Re-added check for >8 outgoing links.
6. Updated some plot colors to look nicer.
7. Added UTF-8 support.
Merged: 2 October 2014


makePlan.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
"""
Ingress Maxfield - makePlan.py

usage: makePlan.py [-h] [-v] [-n NUM_AGENTS] [-s SAMPLES]
                   input_file

Ingress Maxfield - Maximize the number of links and fields, and thus AP, for a
collection of portals in the game Ingress.

positional arguments:
  input_file            Input semi-colon delimited portal file

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         show program's version number and exit
  -n NUM_AGENTS, --num_agents NUM_AGENTS
                        Number of agents. Default: 1
  -s SAMPLES, --samples SAMPLES
                        Number of iterations to perform. More iterations may
                        improve results, but will take longer to process.
                        Default: 50

Original version by jpeterbaker
22 July 2014 - tvw updates csv file format
15 August 2014 - tvw updates with google API, adds -s,
                 switchted to ; delimited file
29 Sept 2014 - tvw V2.0 major update to new version
21 April 2015 - tvw V2.1 force data read to be string
19 Dec 2015 - gbs V3.0 Python 3 compatibility enhancements, color output Resistance Blue
"""

import sys
import os
import argparse
import networkx as nx
import numpy as np
import pandas as pd
from lib import maxfield,PlanPrinterMap,geometry,agentOrder
import pickle

import matplotlib.pyplot as plt


try:
    unicode = unicode
except NameError:
    # 'unicode' is undefined, must be Python 3
    str = str
    unicode = str
    bytes = bytes
    basestring = (str,bytes)
else:
    # 'unicode' exists, must be Python 2
    str = str
    unicode = unicode
    bytes = str
    basestring = basestring

# version number
_V_ = '3.0'
# max portals allowed
_MAX_PORTALS_ = 1000

def main():
    description=("Ingress Maxfield - Maximize the number of links "
                 "and fields, and thus AP, for a collection of "
                 "portals in the game Ingress.")
    parser = argparse.ArgumentParser(description=description,
                                     prog="makePlan.py")
    parser.add_argument('-v','--version',action='version',
                        version="Ingress Maxfield v{0}".format(_V_))
    parser.add_argument('-g','--google',action='store_true',
                        help='Make maps with google maps API. Default: False')
    parser.add_argument('-a','--api_key',default=None,
                        help='Google API key for Google maps. Default: None')
    parser.add_argument('-n','--num_agents',type=int,default='1',
                        help='Number of agents. Default: 1')
    parser.add_argument('-s','--samples',type=int,default=50,
                        help="Number of iterations to "
                        "perform. More iterations may improve "
                        "results, but will take longer to process. "
                        "Default: 50")
    parser.add_argument('input_file',
                        help="Input semi-colon delimited portal file")
    args = vars(parser.parse_args())

    # Number of iterations to complete since last improvement
    EXTRA_SAMPLES = args["samples"]

    GREEN = '#3BF256' # Actual faction text colors in the app
    BLUE  = '#2ABBFF'
    # Use google?
    useGoogle = True
    api_key = args['api_key']

    input_file = args['input_file']

    output_directory = os.path.expanduser('~') + "/Ingress/Fielding/{}".format(os.path.split(args['input_file'])[1][:-4])
    print(output_directory)
    # add ending separator
    if output_directory[-1] != os.sep:
        output_directory += os.sep
    # create directory if doesn't exist
    if not os.path.exists(output_directory):
        os.makedirs(output_directory)
    output_file = (os.path.split(args['input_file'])[1][:-4])
    print(output_file)
    if output_file[-4:] != '.pkl':
        output_file += ".pkl"

    nagents = args["num_agents"]
    if nagents < 0:
        sys.exit("Number of agents should be positive")

    EXTRA_SAMPLES = args["samples"]
    if EXTRA_SAMPLES < 0:
        sys.exit("Number of extra samples should be positive")
    elif EXTRA_SAMPLES > 100:
        sys.exit("Extra samples may not be more than 100")

    if input_file[-3:] != 'pkl':
        # If the input file is a portal list, let's set things up
        a = nx.DiGraph() # network tool
        locs = [] # portal coordinates
        # each line should be name;intel_link;keys
        portals = pd.read_table(input_file,sep=';',
                                comment='#',index_col=False,
                                names=['name','link','keys'],dtype=str)
        portals = np.array(portals)
        portals = np.array([portal for portal in portals if (isinstance(portal[0], basestring) and isinstance(portal[1], basestring))])
        print ("Found {0} portals in portal list.".format(len(portals)))
        if len(portals) < 3:
            sys.exit("Error: Must have more than 2 portals!")
        if len(portals) > _MAX_PORTALS_:
            sys.exit("Error: Portal limit is {0}".\
                     format(_MAX_PORTALS_))
        for num,portal in enumerate(portals):
            if len(portal) < 3:
                print ("Error! Portal ",portal[0]," has a formatting problem.")
                sys.exit()
            a.add_node(num)
            a.node[num]['name'] = portal[0]
            coords = (portal[1].split('pll='))
            if len(coords) < 2:
                print ("Error! Portal ",portal[0]," has a formatting problem.")
                sys.exit()
            coord_parts = coords[1].split(',')
            lat = int(float(coord_parts[0]) * 1.e6)
            lon = int(float(coord_parts[1]) * 1.e6)
            locs.append(np.array([lat,lon],dtype=float))
            try:
                keys = int(portal[2])
                a.node[num]['keys'] = keys
            except ValueError:
                a.node[num]['keys'] = 0

        n = a.order() # number of nodes
        locs = np.array(locs,dtype=float)

        # Convert coords to radians, then to cartesian, then to
        # gnomonic projection
        locs = geometry.e6LLtoRads(locs)
        xyz  = geometry.radstoxyz(locs)
        xy   = geometry.gnomonicProj(locs,xyz)

        for i in range(n):
            a.node[i]['geo'] = locs[i]
            a.node[i]['xyz'] = xyz[i]
            a.node[i]['xy' ] = xy[i]

        # EXTRA_SAMPLES attempts to get graph with few missing keys
        # Try to minimuze TK + 2*MK where
        # TK is the total number of missing keys
        # MK is the maximum number of missing keys for any single
        # portal
        bestgraph = None
        bestlack = np.inf
        bestTK = np.inf
        bestMK = np.inf

        allTK = []
        allMK = []
        allWeights = []

        sinceImprove = 0

        while sinceImprove<EXTRA_SAMPLES:
            b = a.copy()

            sinceImprove += 1

            if not maxfield.maxFields(b):
                print ('Randomization failure\nThe program may work if you try again. It is more likely to work if you remove some portals.')
                continue

            TK = 0
            MK = 0
            for j in range(n):
                keylack = max(b.in_degree(j)-b.node[j]['keys'],0)
                TK += keylack
                if keylack > MK:
                    MK = keylack
            
            weightedlack = TK+2*MK

            allTK.append(TK)
            allMK.append(MK)
            allWeights.append(weightedlack)

            if weightedlack < bestlack:
                sinceImprove = 0
                print ('IMPROVEMENT:\n\ttotal: %s\n\tmax:   %s\n\tweighted: %s'%\
                       (TK,MK,weightedlack))
                bestgraph = b
                bestlack  = weightedlack
                bestTK  = TK
                bestMK  = MK
            else:
                print ('this time:\n\ttotal: %s\n\tmax:   %s\n\tweighted: %s'%\
                       (TK,MK,weightedlack))

            if weightedlack <= 0:
                print ('KEY PERFECTION')
                bestlack  = weightedlack
                bestTK  = TK
                bestMK  = MK
                break
            # if num agent keys is zero, this code isn't true...
            # if all([ b.node[i]['keys'] <= b.out_degree(i) for i in xrange(n) ]):
            #     print 'All keys used. Improvement impossible'
            #     break

            print ('%s tries since improvement'%sinceImprove)

        if bestgraph == None:
            print ('EXITING RANDOMIZATION LOOP WITHOUT SOLUTION!')
            print ('')
            exit()

        print ('Choosing plan requiring %s additional keys, max of %s from single portal'%(bestTK,bestMK))

        plt.clf()
        plt.scatter(allTK,allMK,c=allWeights,marker='o')
        plt.xlim(min(allTK)-1,max(allTK)+1)
        plt.ylim(min(allMK)-1,max(allMK)+1)
        plt.xlabel('Total keys required')
        plt.ylabel('Max keys required for a single portal')
        cbar = plt.colorbar()
        cbar.set_label('Optimization Weighting (lower=better)')
        plt.savefig(output_directory+'optimization.png')

        a = bestgraph

        # Attach to each edge a list of fields that it completes
        # catch no triangulation (bad portal file?)
        try:
            for t in a.triangulation:
                t.markEdgesWithFields()
        except AttributeError:
            print ("Error: problem with bestgraph... no triangulation...?")

        agentOrder.improveEdgeOrder(a)

        with open(output_directory+output_file,'wb') as fout:
            pickle.dump(a,fout)
    else:
        with open(input_file,'r') as fin:
            a = pickle.load(fin)
    #    agentOrder.improveEdgeOrder(a)
    #    with open(output_directory+output_file,'w') as fout:
    #        pickle.dump(a,fout)

    PP = PlanPrinterMap.PlanPrinter(a,output_directory,nagents,color=BLUE,useGoogle=useGoogle,
                                    api_key=api_key)
    PP.keyPrep()
    PP.agentKeys()
    PP.planMap(useGoogle=useGoogle)
    PP.agentLinks()

    # These make step-by-step instructional images
    PP.animate(useGoogle=useGoogle)
    PP.split3instruct(useGoogle=useGoogle)

    print ("Number of portals: {0}".format(PP.num_portals))
    print ("Number of links: {0}".format(PP.num_links))
    print ("Number of fields: {0}".format(PP.num_fields))
    portal_ap = (125*8 + 500 + 250)*PP.num_portals
    link_ap = 313 * PP.num_links
    field_ap = 1250 * PP.num_fields
    print ("AP from portals capture: {0}".format(portal_ap))
    print ("AP from link creation: {0}".format(link_ap))
    print ("AP from field creation: {0}".format(field_ap))
    print ("Total AP: {0}".format(portal_ap+link_ap+field_ap))

if __name__ == "__main__":
    main()
